rfcs/0124-worker-manager.md

```diff
@@ -0,0 +1,271 @@
+alongside one or more staging clusters.
+
+There has also been difficulty in building coherent and reliable worker
+management tools which span more than one cloud.  The Queue is the single
+coordinating system of Taskcluster, but is not an expert at managing workers.
+The systems which *are* experts at worker management do not have a
+programmatically discoverable API.
+
+We would like to change the architecture of Taskcluster to have a new component
+called the worker manager.  This new component would take over the worker
+management responsibilities from the Queue as well as providing a single API to
+manage workers.
+
+The worker manager would be a single instance per cluster.  It would be
+responsible for all facets of worker management.  It would be responsible for:
+
+- storing worker configurations for a cluster
+- storing documentation for workers
+- storing payload schema for workers
+- providing apis to manage workers
+- provisioning through providers and bidding strategies
+- providing credentials to workers as appropriate for their hosting environment
+
+We have a category of 'unprovisioned' workers.  These workers are ones
+which aren't created by an automatic provisioner.  These workers will also need
+the facilities provided by the worker manager, but do not have their capacity
+managed by a system.
+
+## Motivation
+Taskcluster is growing and becoming redeployable.  We would like to be able to
+move into more cloud providers without having to reinvent the wheel each time.
+We would also like managing a cluster to be a pleasant experience.  We would
+like a single, discoverable, reliable, programmatically usable API to manage
+workers.
+
+# Details
+The worker manager will be a single component per cluster.  The worker manager
+will have a single, static, reserved provisionerId of `worker-manager`.  All managed
+instances will use this provisionerId, regardless of whether they are provisioned.
+
+Workers which do not use automated provisioning will be first class citizens in
+the worker manager, but use of the worker manager will not be mandatory.  Tasks
+for other provisionerIds will still be run, but will not have any of the
+management capabilities from the worker manager.  We will build tooling around
+the worker manager and not other systems as a core part of Taskcluster.
+
+## Terminology
+Terminology is often a source of confusion.  This section will clarify the meaning of
+the presented words *in this document*.  It is not intended to be an agreement on general
+usage of the terms outside the scope of this document.
+
+- Provisioner: (old term) a system which provides complete management of a pool
+  of workers from a specific cloud provider.  Example: aws-provisioner
+- Provider: (new term) a system which knows how to create and destroy workers.
+  Does not understand concepts like capacity.  Essentially something which
+  wraps a cloud's instance management APIs.  Similar in scope to existing
+  service ec2-manager
```

Comment:
It would probably be really handy to be able to provision in several accounts in the same cloud, for example to share costs between organizations.  For a collaboratove project like WPT, multiple participants might "donate" capacity by paying the bill for an account (WPT's not a great example since Microsoft and Google would obviously donate capacity from their own clouds!).  I think that just a slight tweak to this proposal could allow that, where providers can be specified with arbitrary names, and just contain configuration for their plugin and relevant account details.  Then we could write `providers: [{name: "ec2-mozilla", plugin: "ec2", account: ..}, {name: "ec2-ieee", plugin: "ec2", account: ..}]`.  Maybe those would be called two "providers" with the same plugin?

I can see projects like the linux kernel or OpenStack, which have several corporate sponsors, wanting features like this.

Reply:
Yes, that's intended.  I've added a little clarification around that, but the intention is that there's provider code and provider instances.  Let me know if the change in language is enough to clarify this.